### ISSUE TITLE
Missing formatting of 'jgrapht-unimi-dsi' and suggestions for improvements of formatting-process

### ISSUE THREAD

**Issue by HannesWell** [jgrapht/jgrapht#1065]
```
 * JGraphT version: 1.5.2-SNAPSHOT
 * Java version (java -version)/platform:  
```

**Issue**

The code of the new `jgrapht-unimi-dsi` module was not formatted at the end of the 1.5.1 release-cycle. This can be fixed by adding a corresponding line in the `codeFormatter.sh`. Alternatively the formatter application could just run in the jgrapht root directory.

Additionally I got the impressions that the squashing of imports was not done.

Nevertheless I want to suggest to use Maven and the Eclipse-Tycho [eclipserun-plugin ](https://www.eclipse.org/tycho/sitedocs/tycho-extras/tycho-eclipserun-plugin/plugin-info.html) to run the formatter application.

This has the following advantages
- no explicit Eclipse installation is required (a minimal Eclipse installation is set-up  by the Tycho plug-in every time it is executed) and therefore no external set-up required, so everybody can run it without the need to prepare anything outside the workspace and the result does not depend on any external tooling.
- platform and computer independent
- easy to update to a new Eclipse version (just modify the corresponding variable in the pom)

I prepared a prototype in my JGraphT fork:
https://github.com/HannesWell/jgrapht/blob/mvnBasedFormatter/etc/codeFormatter.pom

You can run it from the jgrapht root directory with the command: `mvn clean initialize -f etc/codeFormatter.pom`
It runs for about 20sec on my machine (the first time could be a little longer, but Maven caches dependencies).

At the moment the result is a little different compared to the formatting when I run it in my Eclipse-IDE (it seems like the lines are a little longer for the eclipserun-plugin. I suspect an encoding issue). But if you want to continue with the suggested approach, I will investigate the issue.

**Comment by jkinable** [comment]
I indeed forgot to (1) include the jgrapht-unimi-dsi module in the formatter and (2) I also forgot to squash the imports. :(

The eclipse formatting is an absolute pain. In the eclipse 2019 releases, the command line approach to invoke the formatter was broken. Later they fixed it. A  year before that, the formatter rules were no longer backward compatible. First I had to figure out what older version of Eclipse I used, and then manually upgrade to newer versions, exporting and importing the formatting rules along the way. Finally, the current eclipse formatter is mindbogglingly slow (takes about 45 min to run for jgrapht in parallel!). I would love to ditch Eclipse altogether. Aside from my Eclipse rand, it does sound like you found something very useful! I'd be interested to give this a try. 
How does this formatting approach know on what java files to run?

@jsichi Do you foresee any issues?

**Comment by HannesWell** [comment]
Yes Eclipse sometimes has his special characteristics that can be frustrating.
But 45min is extremely long. The suggested approach runs in about 20sec on my computer (the first time it can be a little longer because the eclipse artifacts have to be fetched from Maven-Central, but this is in total maybe one or two minutes depending on your internet connection). But from the `codeFormatter.sh` I assume that you have started a new eclipse instance for every Java file in the repo. I'm actually a bit puzzled because you did not specify a workspace location which is actually required. I cannot run the script on windows but could it be that you spend a long time waiting for many many Eclipse instances to start-ups?

I'm glad you like the idea.
Regarding your question: When the specified The formatter application itself simply runs on all files transitively contained in a directoy that have a registered java source-file extension (which includes `.java` by default), when the specified path is a directory. Usually the last argument specifies that path. At the moment there is just one Java file specified.
Passing the path of the repo's root directory could also be used for the current approach.

So I will proceed to make it ready to use and open a PR once I'm done.

One more question: At the moment you use `Eclipse 2020-03` for formatting?

**Comment by jsichi** [comment]
@jkinable I think at this point I'm supposed to say "what could possibly go wrong!" :)

**Comment by jsichi** [comment]
This seems like it would be a big plus; besides making Joris's life as release manager easier, it would also allow us to run autoformatting in the middle of a release cycle when it made sense (e.g. after inclusion of a large new chunk of code).  And if it allowed us to take care of things like uniform brace insertion via cleanup actions as mentioned in #1052, that's a nice bonus.

**Comment by HannesWell** [comment]
Even tough I personally really like the clean up actions of Eclipse and would appreciate it if some of them are used in the formatting processes, I have to say that it seems to not be feasible at the moment to apply them. The reason is that the Formatter-Application can only apply the formatting rules on java source files but cannot apply cleanUp actions and there is no Eclipse-application to apply Eclipse CleanUps to sources, at least I don't know anyone. Please correct me if I'm wrong.

But if you really would like this to happen, I can have a look if a CleanUp application would be feasible to implement and ask in the corresponding Eclipse-project if they would accept one.
Besides the braces you mentioned such an application could for example also squash and remove unnecessary imports, add missing @Override annotations or remove trailing white-spaces (like described in #1075 )

**Comment by jkinable** [comment]
@HannesWell if you could make the eclipse formatting which we currently run by executing the `codeformatter.sh` automated through, then I would already consider that a big win. Let's focus on that. Once that's working, we could look into potential other additions. Like John suggested, being able to run the formatter after a PR has been merged into the master, would be even better.
